File: src/modifiers/scrollIndicators.ts
import { createModifier } from './createModifier';

/// Visibility of scroll indicators. Maps to SwiftUI's
/// `ScrollIndicatorVisibility`.
export type ScrollIndicatorVisibility =
  | 'automatic'
  | 'visible'
  | 'hidden'
  | 'never';

export interface ScrollIndicatorsParams {
  visibility: ScrollIndicatorVisibility;
}

/// SwiftUI `.scrollIndicators(_:)`. Sets the visibility of scroll
/// indicators within scrollable containers nested in this view. watchOS 9+.
export function scrollIndicators(
  visibility: ScrollIndicatorVisibility
): ReturnType<typeof createModifier>;
export function scrollIndicators(
  params: ScrollIndicatorsParams
): ReturnType<typeof createModifier>;
export function scrollIndicators(
  a: ScrollIndicatorVisibility | ScrollIndicatorsParams
) {
  if (typeof a === 'string') {
    return createModifier('scrollIndicators', { visibility: a });
  }
  return createModifier('scrollIndicators', a);
}

File: apple/Sources/ReactNativeWatchOS/Modifiers/ScrollIndicators.swift
import SwiftUI

/// SwiftUI `.scrollIndicators(_:)`. Sets the visibility of scroll
/// indicators within scrollable containers nested in this view.
/// Available watchOS 9+.
///
/// Note: `ScrollIndicatorVisibility` has a `never` case that the shared
/// `RNWVisibilityParser` (`Visibility`) doesn't model, so the mapping is
/// kept local here.
enum RNWScrollIndicatorsModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("scrollIndicators") { view, params, _ in
            AnyView(view.scrollIndicators(visibility(params.string("visibility"))))
        }
    }

    private static func visibility(_ s: String?) -> ScrollIndicatorVisibility {
        switch s {
        case "visible": return .visible
        case "hidden":  return .hidden
        case "never":   return .never
        default:        return .automatic
        }
    }
}
